Add bulk Redirect import feature

lawrencedine opened the pull request:
Adds a new Feature `Etch.OrchardCore.SEO.Redirects.Import` which allows the user to upload an XLSX file specifying `Title`, `FromUrl` and `ToUrl` and it will create and update all relevant `Redirect` content items.

Uses [FluentExcel](https://github.com/Arch/FluentExcel) to work with Excel.

Adds a controller which shows the form for uploading your redirect file and the download template link accessed via an `AdminMenu` link under the SEO category.

`ImportRedirectsService` is responsible for taking the imported excel rows from the file and turning them into `Redirect` content items and updating existing `Redirects` found by their `FromUrl` field.

Also refactors the validation for `RedirectPart` into a new static class `UrlValidation` and uses that within the `ImportRedirectsService` to validate that the entered Urls are correct. Entries skipped will be  highlighted by row but give no reason currently.

Fixes #16

lawrencedine committed:
Add bulk Redirect import feature

Add a new Feature Etch.OrchardCore.SEO.Redirects.Import which allows the
user to upload an XLSX file specifying Title, FromUrl and ToUrl and it
will create and update all relevant Redirect content items.

Add the relevant Startup file which is gated by the Feature attribute.

Add a new Constants file for storing relevant constants for this feature.

Update Manifest with new feature.

Add references to FluentExcel which is used to work with Excel.

Add a template to the module wwwroot for the user to download.

Add a new controller for ImportRedirects with an Index and Import route.

Add a view for ImportRedirects which shows the form for uploading your
redirect file and the download template link.

Add a new service ImportRedirectsService which is responsible for taking
the imported excel rows from the file and turning them into Redirect
content items and updates to existing Redirects found by their FromUrl
field.

Refactor the validation for RedirectPart into a new static class
UrlValidation and use that within the ImportRedirectsService to validate
that the entered Urls are correct. Entries skipped will be highlighted by
row but give no reason currently.

peterkeating commented on README.md:
```diff
@@ -16,6 +16,10 @@ This module is referencing the beta 3 build of Orchard Core ([`1.0.0-beta3-71077
 
 Define main hostname to redirect all domain variations and force SSL.
 
+#### Import
```
Think this should be within the "Redirects" section, not the "Hostname Redirects" section.

lawrencedine replied:
@peterkeating oops, quite right. I see it also mentions 'hostname redirects'.

Pushed up a commit fixing that.

lawrencedine committed:
Fix error the README

Fix an error in the README. The Import feature description should be under Redirects not Hostname Redirects. Also the description refers to hostname redirects.

peterkeating commented:
@lawrencedine When trying to download the template I'm getting a 404. Looking at the source code I can't see the `ImportRedirectTemplate.xlsx` anywhere.

Also from a UI perspective, it might be better if the download template button was a bit more subtle instead of being a primary button.

peterkeating committed:
Tweak content on import redirects view

- Made the download example template a hyperlink instead of a primary
button
- Changed intro wording to a single sentence that includes a link to
download example template

lawrencedine committed:
Fix missing Redirect import template

Fix the Redirect import template file being missing.

Update .gitignore to allow the wwwroot directory.

lawrencedine commented:
@peterkeating `.gitignore` skipping `wwwroot` from [Etch.OrchardCore.ModuleBoilerplate](https://github.com/EtchUK/Etch.OrchardCore.ModuleBoilerplate) got me again. Have included the template file now.

Agree with making the template download less prominent.

peterkeating committed:
Tweaked casing and added full stops

peterkeating commented:
@lawrencedine Do you think we should be handling updating existing redirects? Currently if I upload the same import template but with different `ToUrl` values, it will create duplicate content items. It does seem to pick up the latest one but that might be coincidence? Don't remember there being any logic to dictate that the last published should be used.

lawrencedine commented:
@peterkeating there is [a check](https://github.com/EtchUK/Etch.OrchardCore.SEO/pull/17/files#diff-5081fadb416a4a3c231e7476b5b08544R70) which is supposed to update existing ones. 

In my testing when I implemented it I did find that this worked, it will create a new version of the Redirect though, maybe that was what you saw?

peterkeating commented:
@lawrencedine Odd, I'll give it another go and debug it at the same time.

peterkeating commented:
@lawrencedine I must have changed the `FromUrl` when I first tested it as trying it again everything works as expected 👍

peterkeating reviewed (approved):
Going to be really useful when migrating an existing site to an Orchard Core implementation, great work @lawrencedine.
